File: _C_satellite/matlab/hw8/satelliteController.m
classdef satelliteController
    % 
    %    This class inherits other controllers in order to organize multiple controllers.
    %
    %----------------------------
    properties
        kp_phi
        kd_phi
        kp_th
        kd_th
        k
    end
    %----------------------------
    methods
        %----------------------------
        function self = satelliteController(P)
            self.kp_phi = P.kp_phi;
            self.kd_phi = P.kd_phi;
            self.kp_th = P.kp_th;
            self.kd_th = P.kd_th;
            self.k = P.k;
        end
        %----------------------------
        function tau = update(self, phi_r, state)
            theta = state(1);
            phi = state(2);
            thetadot = state(3);
            phidot = state(4);
            phi_r = (1+self.kp_phi)/self.kp_phi * phi_r;
            
            % the reference angle for theta comes from the outer loop PD control
            theta_r = self.kp_phi * (phi_r - phi) - self.kd_phi * phidot;
            % the torque applied to the base comes from the inner loop PD control
            tau = self.kp_th * (theta_r - theta) - self.kd_th * thetadot;
        end
    end
end

File: _C_satellite/python/hw8/satelliteController.py
import hw8.satelliteParamHW8 as P

class satelliteController:
    def __init__(self):
        self.kp_phi = P.kp_phi
        self.kd_phi = P.kd_phi
        self.kp_th = P.kp_th
        self.kd_th = P.kd_th

    def update(self, phi_r, state):

        theta = state[0,0]
        phi = state[1,0]
        thetadot = state[2,0]
        phidot = state[3,0]

        # outer loop: outputs the reference angle for theta
        # note that book recommends a feed forward term because
        # of poor DC gain on the outer loop which
        # is why we add an additional "phi_r" at the end
        theta_r = self.kp_phi * (phi_r - phi) \
                  - self.kd_phi * phidot + phi_r

        # inner loop: outputs the torque applied to the base
        tau = self.kp_th * (theta_r - theta) \
              - self.kd_th * thetadot

        return tau
